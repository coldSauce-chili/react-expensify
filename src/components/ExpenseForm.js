import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'


const now = moment()
console.log(now.format('MMM Do, YYYY'))


export default class ExpenseForm extends React.Component {

    // Mead claimed that Props cannot be access without using constructor
    // from video: Section 11 - Wiring Up Edit Expense
    // Debunk below
    constructor(props) {
        super(props)
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ""
        }
    }

    // THIS IS A VALID SYNTAX!!
    //  you can access props by using this.props
    // no need constructor()
    //
    // state = {
    //     description: this.props.expense ? this.props.expense.description : '',
    //     note: this.props.expense ? this.props.expense.note : '',
    //     amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    //     createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    //     calendarFocused: false,
    //     error: ""
    // }




    onDescriptionChange = (e) => {
        const description = e.target.value
        this.setState(() => ({ description }))
        // console.log(this.state.description)
    }
    onNoteChange = (e) => {
        const note = e.target.value
        this.setState(() => ({ note }))
    }
    onAmountChange = (e) => {
        const amount = e.target.value
        if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }))
            // console.log(this.state.amount)

        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }))
        }
    }
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    }
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: "input description and amount" }))
        } else {
            this.setState(() => ({ error: "" }))
            console.log('submitted')
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    render() {
        return (
            <form className='form' onSubmit={this.onSubmit}>
                {this.state.error && <p className='form__error'>{this.state.error}</p>}
                <input
                    type="text"
                    placeholder="description"
                    autoFocus
                    className='text-input'
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="amount"
                    className='text-input'
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => (false)}
                />
                <textarea
                    placeholder="notes"
                    className='textarea'
                    value={this.state.note}
                    onChange={this.onNoteChange}
                >
                </textarea>
                <div>
                    <button className='button'>Add expense</button>
                </div>

            </form>
        )

    }
} 