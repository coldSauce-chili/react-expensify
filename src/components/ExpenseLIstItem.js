//ecxport stateless coponent
// redner description, amout date crated

import React from 'react'
import { Link } from 'react-router-dom'
import moment from 'moment'
import numeral from 'numeral'
const ExpenseListItems = ({ id, description, amount, createdAt }) => (
    <div>
        <Link to={`edit/${id}`}>
            <h4>description: {description}</h4>
        </Link>

        <p>{numeral(amount / 100).format('$0,0.00')} - {moment(createdAt).format('MMMM Do, YYYY')} </p>
    </div>
)
export default ExpenseListItems
