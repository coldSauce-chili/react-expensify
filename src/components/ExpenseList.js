import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseLIstItem'
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div className='content-container'>
        <div className='list-header'>
            <div className='show-for-mobile'>expenses</div>
            <div className='show-for-desktop'>
                expense
            </div>
            <div className='show-for-desktop'>
                amount
            </div>
        </div>
        <div className='list-body'>
            {
                props.expenses.length === 0 ? (

                    <div className='list-item list-item--message'>
                        <span>no expenses</span>
                    </div>
                ) : (
                        props.expenses.map((expense) => (
                            <ExpenseListItem
                                key={expense.id}
                                {...expense}
                            />
                        ))
                    )
            }
        </div>

    </div>
)


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseList)




//the 'below' is how the 'above' works

// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses
//     }
// })(ExpenseList)

// export default ConnectedExpenseList