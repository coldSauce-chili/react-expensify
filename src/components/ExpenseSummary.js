import React from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import selectExpenses from '../selectors/expenses'
import totalExpense from '../selectors/expenses-total'
import numeral from 'numeral'


export const ExpenseSummary = (props) => {

    const expensesCount = props.expenses.length
    const total = numeral(totalExpense(props.expenses) / 100).format('$0,0.00')

    return (
        <div className='page-header'>
            <div className='content-container'>
                <h4 className='page-header__title'>viewing <span>{expensesCount}</span> expenses totalling <span>{total}</span> </h4>
                <div className='page-header__actions'>
                    <Link className='button' to='/create'>Add expense</Link>
                </div>
            </div>
        </div>
    )

}

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}

export default connect(mapStateToProps)(ExpenseSummary)