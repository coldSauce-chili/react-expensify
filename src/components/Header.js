import React from 'react'
import { BrowserRouter, Route, Switch, Link, NavLink } from 'react-router-dom'
import { startLogout } from '../actions/auth'
import { connect } from 'react-redux'
export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink activeClassName="is-active" to="/dashboard" >dash</NavLink>
        <NavLink activeClassName="is-active" to="/create">create</NavLink>
        <NavLink activeClassName="is-active" to="/help">help</NavLink>
        <button onClick={startLogout}>logout</button>
    </header>
)
const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
})
export default connect(undefined, mapDispatchToProps)(Header)