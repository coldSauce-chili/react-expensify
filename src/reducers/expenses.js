
const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [...state, action.expense]

        case 'REMOVE_EXPENSE':
            const filtered = state.filter((expense) => expense.id !== action.id)
            return filtered

        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense
                }
            })

        default:
            return state
    }
}

export default expensesReducer