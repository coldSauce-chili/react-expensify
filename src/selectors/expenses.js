import moment from 'moment'

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {

        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment) : true
        const textMatch = typeof text !== 'string' || expense.description.toLowerCase().includes(text.toLowerCase())


        //before momentjs
        //comparing only by raw epoch time
        //const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate
        //const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate

        return startDateMatch && endDateMatch && textMatch
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1
        }
        else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }
        // else {
        //     return 0
        // }
    })

}

