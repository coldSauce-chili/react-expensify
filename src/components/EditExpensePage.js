import React from 'react'
import { connect } from 'react-redux'
import ExpenseForm from './ExpenseForm'
import { editExpense, startRemoveExpense } from '../actions/expenses'

export class EditExpensePage extends React.Component {

    onRemoveExpense = (expense) => {
        this.props.onRemoveExpense(expense)
        this.props.history.push('/')
    }

    onEditExpense = (id, expense) => {
        this.props.onEditExpense(id, expense)
        this.props.history.push('/')
    }


    render() {
        return (
            <div>
                editing the expens with the ID of {this.props.match.params.id}
                <ExpenseForm
                    expense={this.props.expense}

                    onSubmit={(expense) => {
                        this.onEditExpense(this.props.expense.id, expense)
                    }}
                />
                <button onClick={() => { this.onRemoveExpense(this.props.expense) }}>
                    remove
            </button>

            </div>
        )
    }
}
// const EditExpensePage = (props) => {
//     console.log(props)
//     return (
//         <div>
//             editing the expens with the ID of {props.match.params.id}
//             <ExpenseForm
//                 expense={props.expense}

//                 onSubmit={(expense) => {
//                     props.dispatch(editExpense(props.expense.id, expense))
//                     props.history.push('/')
//                 }}
//             />
//             <button onClick={remove}>remove
//             </button>

//         </div>
//     )
// }


const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
    }
}

const mapDispatchToProps = (dispatch) => {

    return {
        onRemoveExpense: (expense) => {
            dispatch(startRemoveExpense(expense))
            // history.push('/')
        },
        onEditExpense: (id, expense) => {
            dispatch(editExpense(id, expense))
            // history.push('/')
            console.log('this is the update data')
            console.log()
        }
    }

}

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage) 