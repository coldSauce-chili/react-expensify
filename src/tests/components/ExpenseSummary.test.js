import React from 'react'
import { ExpenseSummary } from '../../components/ExpenseSummary'
import { shallow } from 'enzyme'
import expenses from '../fixtures/expenses'

test('should render for 1 item', () => {
    const wrapper = shallow(<ExpenseSummary expenses={[expenses[0]]} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render 0 for no item', () => {
    const wrapper = shallow(<ExpenseSummary expenses={[]} />)
    expect(wrapper).toMatchSnapshot()
})

test('should render multiple item', () => {
    const wrapper = shallow(<ExpenseSummary expenses={expenses} />)
    expect(wrapper).toMatchSnapshot()
})

