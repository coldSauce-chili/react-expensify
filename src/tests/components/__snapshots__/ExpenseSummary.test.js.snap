// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 0 for no item 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h4
      className="page-header__title"
    >
      viewing 
      <span>
        0
      </span>
       expenses totalling 
      <span>
        $0.00
      </span>
       
    </h4>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render for 1 item 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h4
      className="page-header__title"
    >
      viewing 
      <span>
        1
      </span>
       expenses totalling 
      <span>
        $1.95
      </span>
       
    </h4>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render multiple item 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h4
      className="page-header__title"
    >
      viewing 
      <span>
        3
      </span>
       expenses totalling 
      <span>
        $1,141.95
      </span>
       
    </h4>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add expense
      </Link>
    </div>
  </div>
</div>
`;
